source/request-response/subscription_manager.c

```diff
@@ -451,114 +443,150 @@ static int s_rr_activate_idle_subscription(
+        struct aws_rr_subscription_record *existing_record = s_get_subscription_record(manager, topic_filter);
 
-    s_aws_rr_subscription_record_log_invariant_violations(existing_record);
+        if (existing_record == NULL) {
+            existing_record = s_aws_rr_subscription_new(manager->allocator, topic_filter, options->type);
+            aws_hash_table_put(&manager->subscriptions, &existing_record->topic_filter_cursor, existing_record, NULL);
+        }
 
-    // for simplicity, we require unsubscribes to complete before re-subscribing
-    if (existing_record->pending_action == ARRSPAT_UNSUBSCRIBING) {
-        AWS_LOGF_DEBUG(
-            AWS_LS_MQTT_REQUEST_RESPONSE,
-            "request-response subscription manager - acquire subscription for ('" PRInSTR "'), operation %" PRIu64
-            " blocked - existing subscription is unsubscribing",
-            AWS_BYTE_CURSOR_PRI(options->topic_filter),
-            options->operation_id);
-        return AASRT_BLOCKED;
+        s_add_listener_to_subscription_record(existing_record, options->operation_id);
+        if (existing_record->status != ARRSST_SUBSCRIBED) {
+            is_fully_subscribed = false;
+        }
     }
 
-    // register the operation as a listener
-    s_add_listener_to_subscription_record(existing_record, options->operation_id);
-    if (existing_record->status == ARRSST_SUBSCRIBED) {
+    if (is_fully_subscribed) {
         AWS_LOGF_DEBUG(
             AWS_LS_MQTT_REQUEST_RESPONSE,
-            "request-response subscription manager - acquire subscription for ('" PRInSTR "'), operation %" PRIu64
-            " subscribed - existing subscription is active",
-            AWS_BYTE_CURSOR_PRI(options->topic_filter),
+            "request-response subscription manager - acquire subscription for operation %" PRIu64
+            " fully subscribed - all required subscriptions are active",
             options->operation_id);
         return AASRT_SUBSCRIBED;
     }
 
-    // do we need to send a subscribe?
-    if (s_rr_activate_idle_subscription(manager, existing_record)) {
-        // error code was already logged at the point of failure
-        AWS_LOGF_ERROR(
-            AWS_LS_MQTT_REQUEST_RESPONSE,
-            "request-response subscription manager - acquire subscription for ('" PRInSTR "'), operation %" PRIu64
-            " failed - synchronous subscribe failure",
-            AWS_BYTE_CURSOR_PRI(options->topic_filter),
-            options->operation_id);
-        return AASRT_FAILURE;
-    }
+    for (size_t i = 0; i < options->topic_filter_count; ++i) {
+        struct aws_byte_cursor topic_filter = options->topic_filters[i];
+        struct aws_rr_subscription_record *existing_record = s_get_subscription_record(manager, topic_filter);
 
-    s_aws_rr_subscription_record_log_invariant_violations(existing_record);
+        if (s_rr_activate_idle_subscription(manager, existing_record)) {
```

Comment:
If this is failing on the first of a number of topics, does no further action need to be taken on subsequent topics? e.g. we emit a subscription event for the failing record. Should we also emit the subscription event for remaining records or notify that x number of subscription attempts are being skipped due to this one's failure?

Reply:
No, one failure causes the operation to fail which in turn causes the operation to release all its acquisition attempts